src/test/extensions/emulator/optionsHelper.spec.ts

```diff
@@ -0,0 +1,125 @@
+import { expect } from "chai";
+import * as sinon from "sinon";
+
+import * as optionsHelper from "../../../extensions/emulator/optionsHelper";
+import { ExtensionSpec } from "../../../extensions/extensionsApi";
+import * as paramHelper from "../../../extensions/paramHelper";
+
+describe("optionsHelper", () => {
+  describe("getParams", () => {
+    const testOptions = {
+      project: "test",
+      testParams: "test.env",
+    };
+    const autoParams = {
+      PROJECT_ID: "test",
+      EXT_INSTANCE_ID: "test",
+      DATABASE_INSTANCE: "test",
+      DATABASE_URL: "https://test.firebaseio.com",
+      STORAGE_BUCKET: "test.appspot.com",
+    };
+    let testSpec: ExtensionSpec;
+    let readParamsFileStub: sinon.SinonStub;
+
+    beforeEach(() => {
+      testSpec = {
+        name: "test",
+        version: "0.1.0",
+        resources: [],
+        sourceUrl: "https://my.stuff.com",
+        params: [],
+      };
+      readParamsFileStub = sinon.stub(paramHelper, "readParamsFile");
+    });
+
+    afterEach(() => {
+      readParamsFileStub.restore();
+    });
+
+    it("should return user and autopopulated params", () => {
+      testSpec.params = [
+        {
+          label: "param1",
+          param: "USER_PARAM1",
+        },
+        {
+          label: "param2",
+          param: "USER_PARAM2",
+        },
+      ];
+      readParamsFileStub.resolves({
+        USER_PARAM1: "val1",
+        USER_PARAM2: "val2",
+      });
+
+      expect(optionsHelper.getParams(testOptions, testSpec)).to.eventually.deep.eq(
+        Object.assign(
+          {
+            USER_PARAM1: "val1",
+            USER_PARAM2: "val2",
+          },
+          autoParams
+        )
+      );
+    });
+
+    it("should subsitute into params that reference other params", () => {
+      testSpec.params = [
+        {
+          label: "param1",
+          param: "USER_PARAM1",
+        },
+        {
+          label: "param2",
+          param: "USER_PARAM2",
+        },
+        {
+          label: "param3",
+          param: "USER_PARAM3",
+        },
+      ];
+      readParamsFileStub.resolves({
+        USER_PARAM1: "${PROJECT_ID}-hello",
+        USER_PARAM2: "val2",
+        USER_PARAM3: "${USER_PARAM2}",
```

Comment:
I don't think substitution from user param is allowed outside of the emulated environment - it seems like only the auto params are allowed to be substituted https://github.com/firebase/firebase-tools/blob/66447909ab6f2defd04b257b48193c70d778c7ea/src/extensions/askUserForParam.ts#L142.

Can we double check if this is what we want?

Reply:
I looked into the backend, and it seems like we do support user params that reference other user params (see l627 of instance_converter.go). This seems like a pretty big corner case, but I feel like we may as well try to match the backend behavior as closely as posisible. 

askUserForParams is a bit different, since at that point user params don't have any values (we're about to ask for them!), so we only need to substitute in the autpopulated values.

Reply:
Thanks for checking! That's a tricky behavior - my guess is that no user is using this feature, but I agree following the backend behavior is the way to go.